### PR TITLE
[TT-640] Add detailed mode for certificate listing API

### PR DESCRIPTION
**Pull request by jeffy-mathew**
This PR intends to add detailed mode for certificate listing API to return additional details are returned. A query parameter `mode` is added to the API `GET /certs`, when `mode=detailed`, additional details  `issuer.commonName`,  `subject.commonName`, `not_before`, `not_after` are returned

### CODE REVIEW

File: gateway/cert_test.go [hunk 1]
```diff
@@ -1124,6 +1139,18 @@ func TestCertificateHandlerTLS(t *testing.T) {
...
 
+	t.Run("List certificates, detailed mode", func(t *testing.T) {
+		certBasicDetailsTemplate := `{"id":"1%s","issuer_cn":"%s","subject_cn":"%s","not_before":"%s","not_after":"%s"}`
+		clientCertMeta := fmt.Sprintf(certBasicDetailsTemplate, clientCertID, clientCert.Leaf.Issuer.CommonName,
+			clientCert.Leaf.Subject.CommonName, clientCert.Leaf.NotBefore.UTC().Format("2006-01-02T15:04:05Z07:00"), clientCert.Leaf.NotAfter.UTC().Format("2006-01-02T15:04:05Z07:00"))
+		serverCertMeta := fmt.Sprintf(certBasicDetailsTemplate, serverCertID, serverCert.Leaf.Issuer.CommonName,
+			serverCert.Leaf.Subject.CommonName, serverCert.Leaf.NotBefore.UTC().Format("2006-01-02T15:04:05Z07:00"), serverCert.Leaf.NotAfter.UTC().Format("2006-01-02T15:04:05Z07:00"))
+		ts.Run(t, []test.TestCase{
+			{Method: "GET", Path: "/tyk/certs?org_id=1&mode=detailed", AdminAuth: true, Code: 200, BodyMatch: clientCertMeta},
+			{Method: "GET", Path: "/tyk/certs?org_id=1&mode=detailed", AdminAuth: true, Code: 200, BodyMatch: serverCertMeta},
```
Comment (furkansenharputlu) [inline gateway/cert_test.go:1150]: Let's use built-in constants `httpMethod` `http.StatusOk`

File: gateway/cert.go [hunk 2]
```diff
@@ -449,8 +455,18 @@ func (gw *Gateway) certHandler(w http.ResponseWriter, r *http.Request) {
 		if certID == "" {
 			orgID := r.URL.Query().Get("org_id")
 
-			certIds := gw.CertificateManager.ListAllIds(orgID)
-			doJSONWrite(w, http.StatusOK, &APIAllCertificates{certIds})
+			certIDs := gw.CertificateManager.ListAllIds(orgID)
+			mode := r.URL.Query().Get("mode")
```
Comment (furkansenharputlu) [inline gateway/cert.go:459]: It would be better if we put this below of `orgID`

File: gateway/cert.go [hunk 3]
```diff
@@ -449,8 +455,18 @@ func (gw *Gateway) certHandler(w http.ResponseWriter, r *http.Request) {
...
+			if mode == ListDetailed {
+				var certificateBasics = make([]*certs.CertificateBasics, len(certIDs))
+				certificates := gw.CertificateManager.List(certIDs, certs.CertificateAny)
+				for ci, certificate := range certificates {
+					certificateBasics[ci] = certs.ExtractCertificateBasics(certificate, certIDs[ci])
+				}
```
Comment (furkansenharputlu) [inline gateway/cert.go:465]: we can skip a line after closing brackets.

File: gateway/cert.go [hunk 4]
```diff
@@ -449,8 +455,18 @@ func (gw *Gateway) certHandler(w http.ResponseWriter, r *http.Request) {
...
+				doJSONWrite(w, http.StatusOK, &APIAllCertificateBasics{Certs: certificateBasics})
+				return
+			}
```
Comment (furkansenharputlu) [inline gateway/cert.go:468]: same here

File: gateway/cert_test.go [hunk 5]
```diff
@@ -1124,6 +1139,18 @@ func TestCertificateHandlerTLS(t *testing.T) {
...
 	})
 
+	t.Run("List certificates, detailed mode", func(t *testing.T) {
+		certBasicDetailsTemplate := `{"id":"1%s","issuer_cn":"%s","subject_cn":"%s","not_before":"%s","not_after":"%s"}`
+		clientCertMeta := fmt.Sprintf(certBasicDetailsTemplate, clientCertID, clientCert.Leaf.Issuer.CommonName,
+			clientCert.Leaf.Subject.CommonName, clientCert.Leaf.NotBefore.UTC().Format("2006-01-02T15:04:05Z07:00"), clientCert.Leaf.NotAfter.UTC().Format("2006-01-02T15:04:05Z07:00"))
+		serverCertMeta := fmt.Sprintf(certBasicDetailsTemplate, serverCertID, serverCert.Leaf.Issuer.CommonName,
+			serverCert.Leaf.Subject.CommonName, serverCert.Leaf.NotBefore.UTC().Format("2006-01-02T15:04:05Z07:00"), serverCert.Leaf.NotAfter.UTC().Format("2006-01-02T15:04:05Z07:00"))
+		ts.Run(t, []test.TestCase{
+			{Method: "GET", Path: "/tyk/certs?org_id=1&mode=detailed", AdminAuth: true, Code: 200, BodyMatch: clientCertMeta},
```
Comment (furkansenharputlu) [inline gateway/cert_test.go:1149]: Let's use `BodyMatchFunc` to handle both certs at once.

File: gateway/cert_test.go [hunk 6]
```diff
@@ -1124,6 +1139,18 @@ func TestCertificateHandlerTLS(t *testing.T) {
 		}...)
 	})
 
+	t.Run("List certificates, detailed mode", func(t *testing.T) {
+		certBasicDetailsTemplate := `{"id":"1%s","issuer_cn":"%s","subject_cn":"%s","not_before":"%s","not_after":"%s"}`
+		clientCertMeta := fmt.Sprintf(certBasicDetailsTemplate, clientCertID, clientCert.Leaf.Issuer.CommonName,
+			clientCert.Leaf.Subject.CommonName, clientCert.Leaf.NotBefore.UTC().Format("2006-01-02T15:04:05Z07:00"), clientCert.Leaf.NotAfter.UTC().Format("2006-01-02T15:04:05Z07:00"))
+		serverCertMeta := fmt.Sprintf(certBasicDetailsTemplate, serverCertID, serverCert.Leaf.Issuer.CommonName,
+			serverCert.Leaf.Subject.CommonName, serverCert.Leaf.NotBefore.UTC().Format("2006-01-02T15:04:05Z07:00"), serverCert.Leaf.NotAfter.UTC().Format("2006-01-02T15:04:05Z07:00"))
+		ts.Run(t, []test.TestCase{
```
Comment (vaske) [inline gateway/cert_test.go:1148]: we miss here `_, _ =`

File: gateway/cert.go [hunk 7]
```diff
@@ -449,8 +456,19 @@ func (gw *Gateway) certHandler(w http.ResponseWriter, r *http.Request) {
 		if certID == "" {
 			orgID := r.URL.Query().Get("org_id")
```
Comment (furkansenharputlu) [inline gateway/cert.go:457]: Actaully I meant adding `mode` to here as it is just for `GET` requests

Comment (jeffy-mathew) [reply gateway/cert.go:456]: moved it

**Review by vaske — APPROVED** [review]

**Pull request merged**